File: scala/scala-spring-boot/src/test/scala/com/meridal/examples/scalaspringboot/controller/RecordingControllerTest.scala
package com.meridal.examples.scalaspringboot.controller

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.test.web.servlet.MockMvc
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.test.web.servlet.result.MockMvcResultMatchers._
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders._
import org.mockito.Mockito._
import org.hamcrest.Matchers._
import com.fasterxml.jackson.databind.ObjectWriter
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature

import com.meridal.examples.scalaspringboot.model.Recording
import com.meridal.examples.scalaspringboot.service.RecordingService
import org.springframework.http.MediaType

@WebMvcTest
@RunWith(classOf[SpringRunner])
class RecordingControllerTest {

  val noman = new Recording("No-Man", "Together We're Stranger", "NOMANDVD2")
  val bowie = new Recording("David Bowie", "Earthling", "BOWCD10")
  val id = 1234L

  @Autowired
  var mvc: MockMvc = _
  
  @MockBean
  var service: RecordingService = _

  var mapper = new ObjectMapper()
  var jsonWriter: ObjectWriter = _

  @Before
  def setup(): Unit = {
    mapper.configure(SerializationFeature.WRAP_ROOT_VALUE, false)
    jsonWriter = mapper.writer().withDefaultPrettyPrinter()
  }

  @Test
  def testGet(): Unit = {
    when(service.getRecording(id)).thenReturn(noman)
    mvc.perform(get("/api/recording/" + id))
      .andExpect(status().isOk())
      .andExpect(content().string(containsString("No-Man")))
  }

  @Test
  def testGetWithInvalidID(): Unit = {
    when(service.getRecording(id)).thenReturn(null)
    mvc.perform(get("/api/recording/" + id))
      .andExpect(status().isNotFound())
  }

  @Test
  def testGetAll(): Unit = {
    when(service.getAllRecordings()).thenReturn(List(bowie, noman))
    mvc.perform(get("/api/recording/"))
      .andExpect(status().isOk())
      .andExpect(content().string(containsString("No-Man")))
      .andExpect(content().string(containsString("David Bowie")))
  }

  @Test
  def testPost(): Unit = {
     val json = jsonWriter.writeValueAsString(bowie)
     when(service.saveRecording(any(classOf[Recording]))).thenReturn(bowie)
     mvc.perform(post("/api/recording/")
      .contentType(MediaType.APPLICATION_JSON)
      .content(json))
      .andExpect(status().isOk)
      .andExpect(content().contentType(MediaType.APPLICATION_JSON))
      .andExpect(content().string(containsString("David Bowie")))
  }

  @Test
  def testDelete(): Unit = {
    when(service.deleteRecording(id)).thenReturn(noman)
    mvc.perform(delete("/api/recording/" + id))
      .andExpect(status().isOk())
      .andExpect(content().string(containsString("No-Man")))
  }

  @Test
  def testDeleteWithInvalidID(): Unit = {
    when(service.deleteRecording(id)).thenReturn(null)
    mvc.perform(get("/api/recording/" + id))
      .andExpect(status().isNotFound())
  }
}

File: java/spring-boot-simple/src/test/java/com/meridal/examples/springbootmysql/springboot/controller/RecordingControllerTest.java
package com.meridal.examples.springbootmysql.springboot.controller;

import static org.hamcrest.Matchers.*;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.meridal.examples.springbootmysql.springboot.model.Recording;
import com.meridal.examples.springbootmysql.springboot.service.RecordingService;
import org.assertj.core.util.Lists;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import java.util.List;

@WebMvcTest
public class RecordingControllerTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Long ID = 1234L;
    private static final String BOWIE_ARTIST = "David Bowie";
    private static final String BOWIE_TITLE = "Lodger";
    private static final String BOWIE_CAT_NO = "BOWLP1";
    private static final String GENESIS_ARTIST = "Genesis";
    private static final String GENESIS_TITLE = "Three Sides Live";
    private static final String GENESIS_CAT_NO = "GENLP1";

    @Autowired
    private MockMvc mvc;

    @MockBean
    private RecordingService service;

    private ObjectWriter jsonWriter;
    private Recording bowie;
    private Recording genesis;

    @BeforeEach
    public void setup() {
        MAPPER.configure(SerializationFeature.WRAP_ROOT_VALUE, false);
        this.jsonWriter = MAPPER.writer().withDefaultPrettyPrinter();
        this.bowie = new Recording(BOWIE_ARTIST, BOWIE_TITLE, BOWIE_CAT_NO);
        this.genesis = new Recording(GENESIS_ARTIST, GENESIS_TITLE, GENESIS_CAT_NO);
    }

    @Test
    public void testGet() throws Exception {
        when (this.service.getRecording(ID)).thenReturn(this.bowie);
        this.mvc.perform(get("/api/recording/" + ID))
            .andExpect(status().isOk())
            .andExpect(content().string(containsString(BOWIE_ARTIST)))
            .andExpect(content().string(containsString(BOWIE_TITLE)))
            .andExpect(content().string(containsString(BOWIE_CAT_NO)));
    }

    @Test
    public void testGetWithInvalidID() throws Exception {
        when (this.service.getRecording(ID)).thenReturn(null);
        this.mvc.perform(get("/api/recording/" + ID))
            .andExpect(status().isNotFound());
    }

    @Test
    public void testGetAll() throws Exception {
        final List<Recording> recordings = Lists.newArrayList(this.bowie, this.genesis);
        when (this.service.getAllRecordings()).thenReturn(recordings);
        this.mvc.perform(get("/api/recording/"))
            .andExpect(status().isOk())
            .andExpect(content().string(containsString(BOWIE_ARTIST)))
            .andExpect(content().string(containsString(BOWIE_TITLE)))
            .andExpect(content().string(containsString(BOWIE_CAT_NO)))
            .andExpect(content().string(containsString(GENESIS_ARTIST)))
            .andExpect(content().string(containsString(GENESIS_TITLE)))
            .andExpect(content().string(containsString(GENESIS_CAT_NO)));
    }

    @Test
    public void testPost() throws Exception {
        final String json = jsonWriter.writeValueAsString(this.bowie);
        this.bowie.setId(ID);
        when (this.service.saveRecording(any())).thenReturn(this.bowie);
        mvc.perform(post("/api/recording/")
            .contentType(MediaType.APPLICATION_JSON)
            .content(json))
            .andExpect(status().isOk())
            .andExpect(content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(content().string(containsString(BOWIE_ARTIST)))
            .andExpect(content().string(containsString(BOWIE_TITLE)))
            .andExpect(content().string(containsString(BOWIE_CAT_NO)));
    }

    @Test
    public void testPut() throws Exception {
        this.bowie.setId(ID);
        final String json = jsonWriter.writeValueAsString(this.bowie);
        when (this.service.saveRecording(any())).thenReturn(this.bowie);
        mvc.perform(put("/api/recording/" + ID)
            .contentType(MediaType.APPLICATION_JSON)
            .content(json))
            .andExpect(status().isOk())
            .andExpect(content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(content().string(containsString(BOWIE_ARTIST)))
            .andExpect(content().string(containsString(BOWIE_TITLE)))
            .andExpect(content().string(containsString(BOWIE_CAT_NO)));
    }

    @Test
    public void testDelete() throws Exception {
        when (this.service.deleteRecording(ID)).thenReturn(this.bowie);
        this.mvc.perform(delete("/api/recording/" + ID))
            .andExpect(status().isOk())
            .andExpect(content().string(containsString(BOWIE_ARTIST)))
            .andExpect(content().string(containsString(BOWIE_TITLE)))
            .andExpect(content().string(containsString(BOWIE_CAT_NO)));
    }

    @Test
    public void testDeleteWithInvalidID() throws Exception {
        when (this.service.deleteRecording(ID)).thenReturn(null);
        this.mvc.perform(delete("/api/recording/" + ID))
            .andExpect(status().isNotFound());
    }
}